activerecord/lib/active_record/persistence.rb

```diff
@@ -477,11 +477,20 @@ def touch(*names, time: nil)
           changes[column] = write_attribute(column, time)
         end
 
-        changes[self.class.locking_column] = increment_lock if locking_enabled?
-
         clear_attribute_changes(changes.keys)
         primary_key = self.class.primary_key
-        self.class.unscoped.where(primary_key => self[primary_key]).update_all(changes) == 1
+
+        if locking_enabled?
+          lock_column = self.class.locking_column
+          lock_val = send(lock_column)
+          changes[lock_column] = increment_lock
+          if self.class.unscoped.where(primary_key => send(primary_key), lock_column => lock_val).update_all(changes) != 1
+            raise ActiveRecord::StaleObjectError.new(self, "touch")
+          end
+          true
```

Comment:
It's better but still hard to mentally parse. There's a lot of code crammed into these lines. I still feel like we should go with something more along @sgrif's original suggestion:

```ruby
scope = self.class.unscoped.where(primary_key => send(primary_key))

if locking_enabled?
  scope = scope.where(self.class.locking_column => send(lock_column))
  changes[self.class.locking_column] = increment_lock
end

result = scope.update_all(changes) == 1
if !result && locking_enabled?
  raise ActiveRecord::StaleObjectError.new(self, "touch")
end
result
```